airbyte-integrations/connectors/source-gitlab/source_gitlab/components/partition_routers.py

```diff
@@ -0,0 +1,43 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+
+from dataclasses import dataclass
+from typing import Iterable
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.declarative.partition_routers.substream_partition_router import SubstreamPartitionRouter
+from airbyte_cdk.sources.declarative.types import StreamSlice
+
+
+@dataclass
+class GroupStreamsPartitionRouter(SubstreamPartitionRouter):
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        parent_streams = {stream.stream.name: stream.stream for stream in self.parent_stream_configs}
+        groups_list = self.config.get("groups_list")
+        selected_parent = parent_streams["include_descendant_groups"] if groups_list else parent_streams["groups_list"]
+
+        for stream_slice in selected_parent.stream_slices(sync_mode=SyncMode.full_refresh):
+            for record in selected_parent.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream_slice):
+                yield StreamSlice(partition={"id": record["id"]}, cursor_slice={})
+
+
+@dataclass
+class ProjectStreamsPartitionRouter(SubstreamPartitionRouter):
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        parent_stream = self.parent_stream_configs[0].stream
+        projects_list = self.config.get("projects_list", [])
+
+        group_project_ids = set()
+        for stream_slice in parent_stream.stream_slices(sync_mode=SyncMode.full_refresh):
+            for record in parent_stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream_slice):
+                group_project_ids.update({i["path_with_namespace"] for i in record["projects"]})
```

Comment:
We need to use `path_with_namespace` here because this is what users provide in the `projects_list` config section and we should check if the projects we found in groups records are in that list. In addition, namespaces don't change if the user renames a group or a project and the only way to do that is to intentionally change it in settings, which, I believe, should result in full re-sync. What do you think?

Reply:
I think the reasoning where "if the config changes, it should trigger a full refresh" makes sense. I'm fine with this then